ranker/shared/ranker_obj.py

```diff
@@ -162,72 +168,52 @@ def score(self, answer, jaccard_like=False):
-        for subject_index in range(num_nodes):
-            weight_dict_i = []
-            for object_id in range(num_nodes):
-                weight_dict_i.append({})
-            weight_dict.append(weight_dict_i)
-
-        index = {node_id: node_ids.index(node_id) for node_id in node_ids}
+
+        # For each potential edge in the dense answer graph
+        # Make a dictionary of edge source / properties.
+        # For the case where there are redundant edges,
+        # same subject, object, source, property type,
+        # Take the max of the property values
+        # This might happen if two KPs have the same underlying data sources
+        # But for some reason return different publication counts        
+        weight_dict = defaultdict(lambda: defaultdict(\
+            lambda: defaultdict( lambda: defaultdict(float))))
         for edge in edges:
-            subject_id, object_id, edge_weight = (
-                edge["subject"],
-                edge["object"],
-                edge["weight"],
-            )
-            subject_index, object_index = index[subject_id], index[object_id]
-            for edge_source in edge_weight.keys():
-                for edge_property in edge_weight[edge_source].keys():
-                    val = edge_weight[edge_source][edge_property]
-                    if edge_source in weight_dict[subject_index][object_index]:
-                        if (
-                            edge_property
-                            in weight_dict[subject_index][object_index][edge_source]
-                        ):
-                            weight_dict[subject_index][object_index][edge_source][
-                                edge_property
-                            ] = max(
-                                weight_dict[subject_index][object_index][edge_source][
-                                    edge_property
-                                ],
-                                val,
-                            )
-                        else:
-                            weight_dict[subject_index][object_index][edge_source][
-                                edge_property
-                            ] = val
-                    else:
-                        weight_dict[subject_index][object_index][edge_source] = {}
-                        weight_dict[subject_index][object_index][edge_source][
-                            edge_property
-                        ] = val
+            subject = edge["subject"]
+            object = edge["object"]
+            edge_weight = edge["weight"]
+
+            for edge_source, edge_properties in edge_weight.items():
+                for edge_property, edge_val in edge_properties.items():
+                    weight_dict[subject][object][edge_source][edge_property] = \
+                        max(weight_dict[subject][object][edge_source][edge_property], edge_val)
+                    weight_dict[object][subject][edge_source][edge_property] = \
```

Comment:
These two lines actually do the fix. We also max the current property value with the reverse object and subject.